### PR TITLE
Add simulate_rir_ism method for room impulse response simulation

### PR DESCRIPTION
**Pull request by nateanl**
replicate of #2644

### CODE REVIEW

File: torchaudio/prototype/functional/_rir.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    .. properties:: Autograd TorchScript
+
+    Args:
+        room (torch.Tensor): The 1D Tensor to determine the room size. The shape is
+            `(D,)`, where ``D`` is 2 if room is a 2D room, or 3 if room is a 3D room.
+        source (torch.Tensor): The coordinate of the sound source. Tensor with dimensions
+            `(D)`.
+        mic_array (torch.Tensor): The coordinate of microphone array. Tensor with dimensions
+            `(channel, D)`.
+        max_order (int): The maximum number of relfections of the source.
```
Comment (hwangjeff) [inline torchaudio/prototype/functional/_rir.py:141]:
```suggestion
        max_order (int): The maximum number of reflections of the source.
```

File: torchaudio/prototype/functional/_rir.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    .. devices:: CPU
+
+    .. properties:: Autograd TorchScript
+
+    Args:
+        room (torch.Tensor): The 1D Tensor to determine the room size. The shape is
+            `(D,)`, where ``D`` is 2 if room is a 2D room, or 3 if room is a 3D room.
+        source (torch.Tensor): The coordinate of the sound source. Tensor with dimensions
+            `(D)`.
+        mic_array (torch.Tensor): The coordinate of microphone array. Tensor with dimensions
```
Comment (hwangjeff) [inline torchaudio/prototype/functional/_rir.py:139]:
```suggestion
        mic_array (torch.Tensor): Microphone coordinates. Tensor with dimensions
```

File: torchaudio/prototype/functional/_rir.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    .. devices:: CPU
+
+    .. properties:: Autograd TorchScript
+
+    Args:
+        room (torch.Tensor): The 1D Tensor to determine the room size. The shape is
+            `(D,)`, where ``D`` is 2 if room is a 2D room, or 3 if room is a 3D room.
+        source (torch.Tensor): The coordinate of the sound source. Tensor with dimensions
+            `(D)`.
```
Comment (hwangjeff) [inline torchaudio/prototype/functional/_rir.py:138]:
```suggestion
            `(D,)`.
```

File: torchaudio/prototype/functional/_rir.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        mic_array (torch.Tensor): The coordinate of microphone array. Tensor with dimensions
+            `(channel, D)`.
+        max_order (int): The maximum number of relfections of the source.
+        absorption (float or torch.Tensor): The absorption coefficients of wall materials.
+            If the dtype is ``float``, the absorption coefficient is identical to all walls and
```
Comment (hwangjeff) [inline torchaudio/prototype/functional/_rir.py:143]:
```suggestion
            If the dtype is ``float``, the absorption coefficient is identical for all walls and
```

**Review by fakufaku — COMMENTED** [review]
This looks pretty good to me. I have left a few comments about docstrings, and also a suggestion to not support 2D rooms.

File: torchaudio/prototype/functional/_rir.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+
+def _compute_image_sources(
+    room: torch.Tensor,
+    source: torch.Tensor,
+    max_order: int,
+    absorption: torch.Tensor,
+    e_scatter: Optional[torch.Tensor] = None,
+) -> Tuple[Tensor, Tensor]:
+    """Compute image sources in a shoebox-like room.
```
[the remark this answers is not in the document]
Comment (nateanl) [reply torchaudio/prototype/functional/_rir.py:17]: Sure. The `_compute_image_sources` is not public, we can add the citation in `simulate_rir_ism` method along with `pyroomacoustic` citation.

File: torchaudio/prototype/functional/_rir.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    exp_lo = abs(torch.floor((XYZ / 2)))
+    exp_hi = abs(torch.floor((XYZ + 1) / 2))
+    t_lo = tr[:, ::2].unsqueeze(1).repeat(1, XYZ.shape[0], 1)  # (num_band, left walls)
+    t_hi = tr[:, 1::2].unsqueeze(1).repeat(1, XYZ.shape[0], 1)  # (num_band, right walls)
+    att = torch.prod((t_lo**exp_lo) * (t_hi**exp_hi), dim=-1)  # (num_band, num_image_source)
+    return img_loc, att
+
+
+def _hann(x: torch.Tensor, T: int):
+    """Compute the Hann window where the values are truncated based on the value of delay.
```
[the remark this answers is not in the document]
Comment (nateanl) [reply torchaudio/prototype/functional/_rir.py]: addressed in https://github.com/pytorch/audio/pull/2880/commits/826f7c9e096e06c8fa8785e4d8b3159cd757c3d0

File: torchaudio/prototype/functional/_rir.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+            ``absorption`` is a Tensor with dimensions `(num_band, num_wall)`.
+            The shape options are ``[(1, 4), (1, 6), (7, 4), (7, 6)]``.
+            ``num_band`` is `1` if the coefficients is the same for all frequencies, or is `7`
+            if the coefficients are different to different frequencies. `7` refers to the default number
+            of octive bands. (See note in `simulate_rir_ism` method).
+            ``num_wall`` is `4` if the room is a 2D room, representing absorption coefficients
+            of ``"west"``, ``"east"``, ``"south"``, and ``"north"`` walls, respectively.
+            Or it is `6` if the room is a 3D room, representing absorption coefficients
+            of ``"west"``, ``"east"``, ``"south"``, ``"north"``, ``"floor"``, and ``"ceiling"``, respectively.
+        e_scatter (torch.Tensor): The scattering coefficients of wall materials.
```
[the remark this answers is not in the document]
Comment (nateanl) [reply torchaudio/prototype/functional/_rir.py]: Yes, there is no `scatter` argument in image source function. I will provide the argument in the hybrid method.
I have changed both to without `e_` prefix.

File: test/torchaudio_unittest/prototype/functional/functional_test_impl.py [hunk 8]
```diff
@@ -534,3 +539,89 @@ def _debug_plot():
...
+
+    @skipIfNoModule("pyroomacoustics")
+    @parameterized.expand([(2, 1), (3, 4)])
+    def test_simulate_rir_ism_multi_band(self, D, channel):
+        """Test simulate_rir_ism in the case where absorption coefficients are different for all walls."""
+        room_dim = torch.rand(D, dtype=self.dtype, device=self.device) + 5
+        mic_array = torch.rand(channel, D, dtype=self.dtype, device=self.device) + 1
+        source = torch.rand(D, dtype=self.dtype, device=self.device) + 4
+        max_order = 3
+        if D == 2:
```
[the remark this answers is not in the document]
Comment (nateanl) [reply test/torchaudio_unittest/prototype/functional/functional_test_impl.py]: addressed in https://github.com/pytorch/audio/pull/2880/commits/826f7c9e096e06c8fa8785e4d8b3159cd757c3d0

File: torchaudio/prototype/functional/_rir.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+    # multi-band processing
+    if e_abs.shape[0] > 1:
+        if center_frequency is None:
+            center = torch.tensor(
+                [125.0, 250.0, 500.0, 1000.0, 2000.0, 4000.0, 8000.0], dtype=room.dtype, device=room.device
+            )
+        else:
+            center = center_frequency
+        filters = torch.ops.torchaudio.make_rir_filter(center, sample_rate, 512)
```
[the remark this answers is not in the document]
Comment (nateanl) [reply torchaudio/prototype/functional/_rir.py]: It is set as default value for sake of simplicity. I have made a comment above the code.

File: torchaudio/prototype/functional/_rir.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    irs = img_src_att[..., None] * _frac_delay(delay, delay_i, delay_filter_length)[None, ...]
+
+    rir_length = int(delay_i.max() + irs.shape[-1])
+    rir = torch.ops.torchaudio.build_rir(irs, delay_i.type(torch.int32), rir_length)
+
+    # multi-band processing
+    if e_abs.shape[0] > 1:
+        if center_frequency is None:
+            center = torch.tensor(
+                [125.0, 250.0, 500.0, 1000.0, 2000.0, 4000.0, 8000.0], dtype=room.dtype, device=room.device
```
[the remark this answers is not in the document]
Comment (nateanl) [reply torchaudio/prototype/functional/_rir.py:247]: It is mentioned in the note of docstring.

File: docs/source/prototype.functional.rst [hunk 11]
```diff
@@ -22,3 +22,8 @@ DSP
    oscillator_bank
    sinc_impulse_response
    frequency_impulse_response
+
+simulate_rir_ism
+~~~~~~~~~~~~~~~~
+
+.. autofunction:: simulate_rir_ism
```
Comment (mthrok) [inline docs/source/prototype.functional.rst:29]: Can you add a new section and use autosummary?

File: torchaudio/_extension/__init__.py [hunk 12]
```diff
@@ -72,6 +73,15 @@
         _LG.debug("Failed to initialize ffmpeg bindings", exc_info=True)
 
 
+# Initialize rir-related features
+_IS_RIR_AVAILABLE = False
+if _IS_TORCHAUDIO_EXT_AVAILABLE:
+    _load_lib("libtorchaudio")
+
+    import torchaudio.lib._torchaudio  # noqa
+
```
Comment (mthrok) [inline torchaudio/_extension/__init__.py:82]: This part is now repeated multiple times, which is redundant. Can you merge the initialization logics around extension, kaldi, RIR as follow?

```
_IS_RIR_AVAILABLE = False
_IS_KALDI_AVAILABLE = False
_IS_TORCHAUDIO_EXT_AVAILABLE = is_module_available("torchaudio.lib._torchaudio")
if _IS_TORCHAUDIO_EXT_AVAILABLE:
    _load_lib("libtorchaudio")

    import torchaudio.lib._torchaudio # noqa

    _check_cuda_version()

    _IS_RIR_AVAILABLE = torchaudio.lib._torchaudio.is_rir_available()
    _IS_KALDI_AVAILABLE = torchaudio.lib._torchaudio.is_kaldi_available()
```

**Pull request closed without merging**